### ISSUE TITLE
Metadata Merge II: how do we define a metadata merge?

### ISSUE THREAD

**Issue by iherman** [w3c/csvw#77]
(Following up on the example in issue #76)

In general, merging objects "A" and "B" is simple when it comes to simple k/v pairs: each key in "B" is copied to "A", possibly overwriting what is there. One can also define that to be "deep", i.e., if the value is an object and the same key also appears in "A" then, instead of simply performing a replacement, a (recursive) merge is performed on the respective value objects and the result will be in the combined object. That seems to work well for us.

However... how do we define the merges when we have arrays? Do we 
- Concatenate the arrays? or 
- Take each element of the array with the same index and perform a recursive merge if the are objects? or 
- Ignore arrays and treat them as simple entries, i.e., we simply copy (replace) from "B" to "A"

The example in issue #76 suggests that element-by-element merge is necessary for the array in "columns". Indeed if we follow the third option, that means the column names, as defined in the CSV file itself, is lost (although that maybe what we want!)

However, there may be other cases when we want option one, i.e., concatenation; consider the list of keywords.

In any case, what "merging" means must be specified, the current document is underspecified. And it would be good to have some general rule and _not_ going down the road of "if the key is 'foo', then we do it this way, otherwise that way..." (some sort of a "semantic merge"). Doing this may lead to a complex specification and an even more complex implementation...

**Comment by gkellogg** [comment]
The issues here might indicate that we only support a shallow merge, so only top-level key values are replaced, rather than the complexity of a deep merge.

If we do a deep merge, then if should probably be recursive, meaning that there are even more issues of comparable values at each level don't have the same form.

Merge scalar or object into array => append to array
Merge array into scalar or object => prepend sclar/object to array
Merge scalar/object into scalar/object => array of values
Merge array into array => many complex options.

Really, shallow is much easier.
Merge

**Comment by iherman** [comment]
While I am sympathetic to this, obviously, it does not solve the issue #78: what that would mean is that the merged metadata would have duplicate information on a given column (one as a top-level "scheme" structure, and the other as part of the "resources" structure...)

Ivan

> On 28 Nov 2014, at 15:52 , Gregg Kellogg notifications@github.com wrote:
> 
> The issues here might indicate that we only support a shallow merge, so only top-level key values are replaced, rather than the complexity of a deep merge.
> 
> If we do a deep merge, then if should probably be recursive, meaning that there are even more issues of comparable values at each level don't have the same form.
> 
> Merge scalar or object into array => append to array
> Merge array into scalar or object => prepend sclar/object to array
> Merge scalar/object into scalar/object => array of values
> Merge array into array => many complex options.
> 
> Really, shallow is much easier.
> Merge
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/w3c/csvw/issues/77#issuecomment-64901461.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by iherman** [comment]
Admin: This is an answer to [Jeni's comment](https://github.com/w3c/csvw/issues/76#issuecomment-64967276) given on issue #76, but the discussion is more relevant on this thread...

Jeni, 

what you describe is what I referred to as a "semantic" merge (lack of a better word), i.e., not some sort of a mechanical merge of (JSON) objects (like what jQuery or other tools do for JavaScript) but, rather, an enrichment of the metadata assigned to a cell, and the way that enrichment happens will depend on the specific metadata. Ie, you say a "title" can be an array, so the enrichment/merge means creating that array, whereas the merge of the 'datatype' term means a direct replacement. 

We may have to go down that route, but that means for each individual term we have to have a clear idea what the merge means. For example, the external templates may be arrays, and I do not know how the merge works. (Maybe, as @gkellogg  proposed, we do a mechanical replacement on the top level for everything that is not the 'scheme'? But that may kill annotations, for example...) We also have to define what happens if the metadata include terms that we do not control (e.g., additional Dublin Core or schema.org terms). I am still hoping (maybe naïvely) that we can have a general procedure that would work for us, but I may be wrong.

(In any case, yes, lots of examples will be needed for the metadata merge. Luckily, we can make it a bit simpler, I guess, because we do not have separate metadata entries defined for rows any more, do we?)

**Comment by 6a6d74** [comment]
@iherman: for metadata properties we don't control I would expect to treat each property like an RDF triple and "just add it to the graph" - so for example if I had, say the property 'dct:license' provided in multiple metadata files, the mapped output would include _all_ those properties.

So a (plain old) JSON mapping might say:

```
"dct:license": "http://www.nationalarchives.gov.uk/doc/open-government-licence/version/2/"
"dct:license": "http://www.nationalarchives.gov.uk/doc/open-government-licence/version/2/"
```

(note that in plain old JSON we're just treating the prefix+propertyname as a simple property name, ugly as it is)

Of course, we might want to consider grouping duplicate properties into an array (given that most JavaScript implementations will ignore all but the last value)?

For RDF mappings, _implementations_ may want to add metadata from different metadata-docs into different Graphs so that the provenance of those statements is maintained ... 

If we're going to express this Graph based approach in the RDF mapping then I'm going to need a little help with Graph syntax(es).

**Comment by iherman** [comment]
> On 01 Dec 2014, at 11:24 , Jeremy Tandy notifications@github.com wrote:
> 
> @iherman: for metadata properties we don't control I would expect to treat each property like an RDF triple and "just add it to the graph" - so for example if I had, say the property 'dct:license' provided in multiple metadata files, the mapped output would include all those properties.
> 
> So a (plain old) JSON mapping might say:
> 
> "dct:license": "http://www.nationalarchives.gov.uk/doc/open-government-licence/version/2/"
> "dct:license": "http://www.nationalarchives.gov.uk/doc/open-government-licence/version/2/"
> 
> (note that in plain old JSON we're just treating the prefix+propertyname as a simple property name, ugly as it is)
> 
> Of course, we might want to consider grouping duplicate properties into an array (given that most JavaScript implementations will ignore all but the last value)?
> 
> For RDF mappings, implementations may want to add metadata from different metadata-docs into different Graphs so that the provenance of those statements is maintained ...

Let us not go there... Just as for Json, we can simply repeat the triples that come from the different metadata files.

> If we're going to express this Graph based approach in the RDF mapping then I'm going to need a little help with Graph syntax(es).

Right. What bothers me is the general issue: we are heading towards a fairly complex way of defining the merge of our metadata, very much depending on the metadata property. In some cases we override, in some cases we repeat, in some cases we create arrays... this is getting messy:-(

Ivan

> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by gkellogg** [comment]
Resolved in PR #169 and on http://www.w3.org/2015/01/28-csvw-minutes.html.

**Comment by gkellogg** [comment]
RESOLUTION: We will resolve #43, #76, #77, #78 as suggested in https://lists.w3.org/Archives/Public/public-csv-wg/2015Jan/0050.html
